File: services/go-gin/internal/sample/backend.go
package sample

import (
	"bytes"
	"encoding/json"
	"github.com/gin-gonic/gin"
	"net/http"
	"sample-micro-service/pkg/service/config"
	"strconv"
	"time"
)

type Opera struct {
	Id           string     `json:"id"`
	Name         string     `json:"name"`
	Composer     string     `json:"composer"`
	ComposedAt   time.Time  `json:"composed_at"`
	PublishedAt  *time.Time `json:"published_at"`
	Description  *string    `json:"description"`
	NumberOfActs int        `json:"number_of_acts"`
	Style        *string    `json:"style"`
	OpenAir      bool       `json:"open_air"`
}

type Planet struct {
	Id           string    `json:"id"`
	Name         string    `json:"name"`
	Diameter     int       `json:"diameter"`
	Orbit        int       `json:"orbit"`
	Gas          bool      `json:"gas"`
	DiscoveredAt time.Time `json:"discovered_at"`
	DiscoveredBy string    `json:"discovered_by"`
	History      *string   `json:"history"`
	Missions     []string  `json:"missions"`
	Moons        []Moon    `json:"moons"`
}

type Moon struct {
	Name         string     `json:"name"`
	Diameter     int        `json:"diameter"`
	Distance     int        `json:"distance"`
	DiscoveredAt *time.Time `json:"discovered_at"`
	DiscoveredBy *string    `json:"discovered_by"`
	PossibleLife bool       `json:"possible_life"`
}

type Journal struct {
	Id          string     `json:"id"`
	Name        string     `json:"name"`
	Title       string     `json:"title"`
	Issue       int        `json:"issue"`
	Publisher   *string    `json:"publisher"`
	PublishedAt *time.Time `json:"published_at"`
	Editors     []string   `json:"editors"`
	Url         *string    `json:"url"`
	Articles    []Article  `json:"articles"`
}

type Article struct {
	Title         string     `json:"title"`
	Description   string     `json:"description"`
	Authors       []string   `json:"authors"`
	Keywords      []string   `json:"keywords"`
	FromPage      *int       `json:"from_page"`
	ToPage        *int       `json:"to_page"`
	LastUpdatedAt *time.Time `json:"last_updated_at"`
	Sections      []Section  `json:"sections"`
}

type Section struct {
	Title         string     `json:"title"`
	Summary       string     `json:"summary"`
	Words         int        `json:"words"`
	LastUpdatedAt *time.Time `json:"last_updated_at"`
}

var (
	fetchUrl    = config.GetString("backend.fetch.url") + "/api/fetch"
	pushUrl     = config.GetString("backend.push.url") + "/api/push"
	httpTimeout = time.Duration(config.GetOptionalInt("backend.timeout-ms", 5_000)) * time.Millisecond

	httpTransport = &http.Transport{
		MaxIdleConns:       10,
		IdleConnTimeout:    30 * time.Second,
		DisableCompression: true,
	}
	httpClient = &http.Client{
		Transport: httpTransport,
		Timeout:   httpTimeout,
	}
)

func fetchSmall(c *gin.Context, userToken string, id string) (Opera, bool) {
	return getRequest(c, fetchUrl+"/"+id+"/small", userToken, Opera{})
}

func fetchMedium(c *gin.Context, userToken string, id string) (Planet, bool) {
	return getRequest(c, fetchUrl+"/"+id+"/medium", userToken, Planet{})
}

func fetchLarge(c *gin.Context, userToken string, id string) (Journal, bool) {
	return getRequest(c, fetchUrl+"/"+id+"/large", userToken, Journal{})
}

func pushSmall(c *gin.Context, userToken string, id string, body Opera) (Opera, bool) {
	return postRequest(c, pushUrl+"/"+id+"/small", userToken, body, Opera{})
}

func pushMedium(c *gin.Context, userToken string, id string, body Planet) (Planet, bool) {
	return postRequest(c, pushUrl+"/"+id+"/medium", userToken, body, Planet{})
}

func pushLarge(c *gin.Context, userToken string, id string, body Journal) (Journal, bool) {
	return postRequest(c, pushUrl+"/"+id+"/large", userToken, body, Journal{})
}

func getRequest[R any](c *gin.Context, url string, userToken string, defaultResult R) (R, bool) {
	req, err := http.NewRequestWithContext(c, "GET", url, nil)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Authorization", "Bearer "+userToken)

	resp, err := httpClient.Do(req)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		c.JSON(http.StatusBadGateway, Error{Message: strconv.Itoa(resp.StatusCode)})
		return defaultResult, false
	}

	var parsedResponse R
	err = json.NewDecoder(resp.Body).Decode(&parsedResponse)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}

	return parsedResponse, true
}

func postRequest[R any](c *gin.Context, url string, userToken string, body R, defaultResult R) (R, bool) {
	serializedBody := new(bytes.Buffer)
	err := json.NewEncoder(serializedBody).Encode(body)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}
	req, err := http.NewRequestWithContext(c, "POST", url, serializedBody)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+userToken)

	resp, err := httpClient.Do(req)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		c.JSON(http.StatusBadGateway, Error{Message: strconv.Itoa(resp.StatusCode)})
		return defaultResult, false
	}

	var parsedResponse R
	err = json.NewDecoder(resp.Body).Decode(&parsedResponse)
	if err != nil {
		c.JSON(http.StatusInternalServerError, Error{Message: err.Error()})
		return defaultResult, false
	}

	return parsedResponse, true
}

File: services/rust-actix/src/sample/backend.rs
use std::time::Duration;

use actix_web::HttpResponse;
use chrono::{DateTime, Utc};
use reqwest::{Client, Response};
use serde::{Deserialize, Serialize};

use crate::sample::error::ErrorResponse;

#[derive(Serialize, Deserialize)]
pub struct Opera {
    pub id: String,
    pub name: String,
    pub composer: String,
    pub composed_at: DateTime<Utc>,
    pub published_at: Option<DateTime<Utc>>,
    pub description: Option<String>,
    pub number_of_acts: u64,
    pub style: Option<String>,
    pub open_air: bool,
}

#[derive(Serialize, Deserialize)]
pub struct Planet {
    pub id: String,
    pub name: String,
    pub diameter: u64,
    pub orbit: u64,
    pub gas: bool,
    pub discovered_at: DateTime<Utc>,
    pub discovered_by: String,
    pub history: Option<String>,
    pub missions: Option<Vec<String>>,
    pub moons: Vec<Moon>,
}

#[derive(Serialize, Deserialize)]
pub struct Moon {
    pub name: String,
    pub diameter: u64,
    pub distance: u64,
    pub discovered_at: Option<DateTime<Utc>>,
    pub discovered_by: Option<String>,
    pub possible_life: bool,
}

#[derive(Serialize, Deserialize)]
pub struct Journal {
    pub id: String,
    pub name: String,
    pub title: String,
    pub issue: u64,
    pub publisher: Option<String>,
    pub published_at: Option<DateTime<Utc>>,
    pub editors: Option<Vec<String>>,
    pub url: Option<String>,
    pub articles: Vec<Article>,
}

#[derive(Serialize, Deserialize)]
pub struct Article {
    pub title: String,
    pub description: String,
    pub authors: Vec<String>,
    pub keywords: Option<Vec<String>>,
    pub from_page: Option<u64>,
    pub to_page: Option<u64>,
    pub last_updated_at: Option<DateTime<Utc>>,
    pub sections: Option<Vec<Section>>,
}

#[derive(Serialize, Deserialize)]
pub struct Section {
    pub title: String,
    pub summary: String,
    pub words: u64,
    pub last_updated_at: Option<DateTime<Utc>>,
}

#[derive(Clone)]
pub struct BackendClient {
    inner: Client,
}

pub fn build_backend_client() -> reqwest::Result<BackendClient> {
    Client::builder()
        .timeout(Duration::from_millis(2_000))
        .build()
        .map(|c| BackendClient { inner: c })
}

impl BackendClient {
    pub async fn fetch_small(&self, user_token: &str, id: &str) -> Result<Opera, HttpResponse> {
        self.fetch("small", user_token, id).await
    }

    pub async fn fetch_medium(&self, user_token: &str, id: &str) -> Result<Planet, HttpResponse> {
        self.fetch("medium", user_token, id).await
    }

    pub async fn fetch_large(&self, user_token: &str, id: &str) -> Result<Journal, HttpResponse> {
        self.fetch("large", user_token, id).await
    }

    pub async fn push_small(&self, user_token: &str, id: &str, body: Opera) -> Result<Opera, HttpResponse> {
        self.push("small", user_token, id, body).await
    }

    pub async fn push_medium(&self, user_token: &str, id: &str, body: Planet) -> Result<Planet, HttpResponse> {
        self.push("medium", user_token, id, body).await
    }

    pub async fn push_large(&self, user_token: &str, id: &str, body: Journal) -> Result<Journal, HttpResponse> {
        self.push("large", user_token, id, body).await
    }

    async fn fetch<T: for<'a> Deserialize<'a>>(&self, path: &str, user_token: &str, id: &str) -> Result<T, HttpResponse> {
        let base_url = std::env::var("BACKEND_FETCH_URL")
            .expect("BACKEND_FETCH_URL must be set");

        BackendClient::wrap_response(self.inner.get(format!("{}/api/fetch/{}/{}", base_url, id, path))
            .bearer_auth(user_token)
            .send()
            .await).await
    }

    async fn push<T: for<'a> Deserialize<'a>, U: Serialize>(&self, path: &str, user_token: &str, id: &str, body: U) -> Result<T, HttpResponse> {
        let base_url = std::env::var("BACKEND_PUSH_URL")
            .expect("BACKEND_PUSH_URL must be set");

        BackendClient::wrap_response(self.inner.post(format!("{}/api/push/{}/{}", base_url, id, path))
            .bearer_auth(user_token)
            .json(&body)
            .send()
            .await).await
    }

    async fn wrap_response<T: for<'a> Deserialize<'a>>(result: reqwest::Result<Response>) -> Result<T, HttpResponse> {
        if let Err(e) = result {
            return Err(HttpResponse::BadGateway().json(ErrorResponse { message: e.to_string() }));
        }
        let response = result.unwrap();

        if response.status() != 200 {
            return Err(HttpResponse::BadGateway().json(ErrorResponse { message: response.status().to_string() }));
        }

        match response.json::<T>().await {
            Err(e) => Err(HttpResponse::BadGateway().json(ErrorResponse { message: e.to_string() })),
            Ok(body) => Ok(body),
        }
    }
}
